src/io/inputs/model-converter/include/antares/io/inputs/model-converter/ForbiddenNodes.h

```diff
@@ -0,0 +1,213 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Mozilla Public Licence 2.0 for more details.
+ *
+ * You should have received a copy of the Mozilla Public Licence 2.0
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#pragma once
+
+#include <stdexcept>
+#include <typeindex>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "antares/expressions/nodes/FunctionNode.h"
+#include "antares/expressions/nodes/Node.h"
+
+namespace Antares::IO::Inputs::ModelConverter
+{
+template<typename NodeType>
+std::type_index forbiddenNodeKey()
+{
+    if constexpr (std::is_same_v<NodeType, Expressions::Nodes::FunctionNode>)
+    {
+        static_assert(!std::is_same_v<NodeType, Expressions::Nodes::FunctionNode>,
+                      "Use Expressions::Nodes::FunctionNodeType enum values or "
+                      "forbiddenNodeKey(Expressions::Nodes::FunctionNodeType) "
+                      "instead of FunctionNode for forbidden rules.");
+    }
+    return std::type_index(typeid(NodeType));
+}
+
+template<Expressions::Nodes::FunctionNodeType T>
+inline std::type_index forbiddenNodeKey()
+{
+    using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType, T>;
+    return std::type_index(typeid(Tag));
+}
+
+inline std::type_index forbiddenNodeKey(const Expressions::Nodes::FunctionNodeType& funcType)
+{
+    switch (funcType)
+    {
+    case Expressions::Nodes::FunctionNodeType::max:
+    {
+        using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType,
+                                           Expressions::Nodes::FunctionNodeType::max>;
+        return std::type_index(typeid(Tag));
+    }
+    case Expressions::Nodes::FunctionNodeType::min:
+    {
+        using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType,
+                                           Expressions::Nodes::FunctionNodeType::min>;
+        return std::type_index(typeid(Tag));
+    }
+    case Expressions::Nodes::FunctionNodeType::pow:
+    {
+        using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType,
+                                           Expressions::Nodes::FunctionNodeType::pow>;
+        return std::type_index(typeid(Tag));
+    }
+    case Expressions::Nodes::FunctionNodeType::dual:
+    {
+        using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType,
+                                           Expressions::Nodes::FunctionNodeType::dual>;
+        return std::type_index(typeid(Tag));
+    }
+    case Expressions::Nodes::FunctionNodeType::reduced_cost:
+    {
+        using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType,
+                                           Expressions::Nodes::FunctionNodeType::reduced_cost>;
+        return std::type_index(typeid(Tag));
+    }
```

Comment:
It seems like the switch..case is not needed here

```suggestion
using Tag = std::integral_constant<Expressions::Nodes::FunctionNodeType, funcType>;
return std::type_index(typeid(Tag));
```

Reply:
you can't do this. funcType is a runtime arg